### ISSUE TITLE
Lecture 4. Optimization

### ISSUE THREAD

**Issue by seonghunn** [seonghunn/CS182#4]
# Part 1
## How does gradient descent work?

<img width="609" alt="image" src="https://user-images.githubusercontent.com/69765278/159139117-ba02ba47-d520-4705-ad1b-eb52310a4a18.png">

공이 중력에 의해 아래로 떨어지듯이, 어디로 떨어질지에 대한 방향을 gradient descent 를 통해 정해주는 것
이를 small constant (learning rate , step size) 를 더해주는 방법으로 사용

<img width="554" alt="image" src="https://user-images.githubusercontent.com/69765278/159139167-45bc1fdd-e07e-438e-a93f-ae3de1b9817c.png">

기울기의 **반대** 방향으로 진행하도록 해야 하기 때문에, partial derivative 에 음의 부호를 붙여 주어야 한다.
contour : 등고선의 한 선과 같이 동일한 level에 해당하는 선이라고 생각할 수 있음 
공이 가장 낮은 곳을 향해 굴러가듯이 적절한 Learning rate를 사용하여 최적해를 찾는데에 사용할 수 있다.

하지만 **항상 최적해를 찾는 것은 아니다** 라는 문제가 발생한다.

<img width="390" alt="image" src="https://user-images.githubusercontent.com/69765278/159139349-809c4cd7-aa18-4dd1-93e4-ec75a783481e.png">

만약에 loss surface 가 다음과 같이 생겼다면

<img width="297" alt="image" src="https://user-images.githubusercontent.com/69765278/159139391-4d1107bb-4d18-440d-b53d-2a9ff9a5ff86.png">

gradient descent를 통해 항상 최적해를 찾아낼 수 있다. 이는 loss가 convex 하기 때문이다.

<img width="273" alt="image" src="https://user-images.githubusercontent.com/69765278/159139425-0c770004-1ded-45c4-8421-65307534649f.png">

logistic regressing 에서의 NLL은 convex가 보장되어 있다. (항상 그런것은 아님, 대부분 사실)
convex 란, 함수의 두 점을 이은 선이 함수보다 항상 위에 있는 것을 말한다 (아래로 볼록)
convex 함수에서는 gradient descent 가 최적해를 찾을 가능성이 매우 높다. 하지만 모든 convex function에서 최적해를 찾을 수 있는 것은 아니다.

NN에서는 굉장히 많은 parameter가 있기 때문에 단순히 convex 한 loss landscape는 나오지 않을 가능성이 높다. 아래와 같은 세 가지 특성이 gradient descent를 사용하는 것을 방해한다.

<img width="652" alt="image" src="https://user-images.githubusercontent.com/69765278/159139618-45df6ff6-4b59-4fce-a554-b6cc639737eb.png">

### Local optima
loss landscape에서 가장 많이 발생하는 이슈이다. gradient descent를 사용했을 때 최적해가 아닌 local solution으로 수렴해 버릴 수 있다.
하지만 parameter를 증가시킬수록 local optima가 점점 줄어들고 있다.

### Plateaus
기울기가 굉장히 적은 곳이다 (평지). 이 존재로 인해 learning rate를 너무 적게하면 그저 진동하며 갖혀버리기 때문에 마냥 작게할 수 없다. 이를 해결하기 위해 **Momentum** 을 활용한다.

### Saddle point
안장점, 한 방향에서는 극대지만 다른 방향에서는 극소인 경우이다. NN loss landscape은 parameter가 굉장히 많기 때문에 대부분의 critical point(극대,극소,안장점)은 대부분 안장점이다.
critical point 는 gradient가 0인 모든 점이다 (극대,극소,안장점)
두 번 미분을 통해 같은 critical point라도 극대인지, 극소인지를 알아낼 수 있고 두 번 미분한 것을 matrix에 배치한 hessian matrix가 있는데
**saddle 에서는 Hessian 의 diagonal entries가 양수 음수가 섞여있어야 한다** 극대는 모든 diagonal entries가 양수, 극소는 모두 음수이다.

# Part 2
## Newton's method
**Newton's method 를 사용** 하여 해결할 수 있다. 이는 감당하기 힘들어 잘 사용하진 않지만, 이상적인 해결 방법이다.
식을 2차까지 테일러 전개하는 방식을 응용하여 사용한다.

<img width="429" alt="image" src="https://user-images.githubusercontent.com/69765278/159140072-96e5ad32-2428-4d76-8d6b-2775efbb7d4f.png">

마지막 줄에 있듯이, 이를 풀기 위해서 Hessian의 역행렬을 구해야 한다.
Newton mtd를 사용하지 못하는 이유가 이 때문인데, 이 계산 과정 때문에 런타임이 굉장히 오래 걸리기 때문이다.

<img width="661" alt="image" src="https://user-images.githubusercontent.com/69765278/159140188-def1f48b-2f32-440b-97ba-ac500f57249e.png">

n개의 parameter가 있다면 gradient descent 를 사용하는 데에 O(n) 만큼의 시간이 걸리고,
n by n matrix인 Hessian의 역행렬을 구하는 데에 O(n<sup>3</sup>)의 시간이 걸려 굉장히 많은 시간이 소요된다.

## Momentum
더 빨리 최적화를 하기 위한 방법으로, 연속적인 gradient step을 평균내어 진행방향의 속도를 결정하는 방법으로 응용하였다.

<img width="556" alt="image" src="https://user-images.githubusercontent.com/69765278/159140240-20142e0e-4b98-4ec2-83cb-0d6e99755efd.png">

기본적인 개요는 연속적인 gradient step이 다른 방향이라면, 그 방향으로의 진행을 그만큼 cancel 하고 연속적인 방향이 같은 방향이라면 더 빠르게 그 방향으로 진행하는 것이다.

<img width="569" alt="image" src="https://user-images.githubusercontent.com/69765278/159140317-d80b290e-c4b4-42f9-aec8-62f9784da15a.png">

이전의 gradient descent와 유사한 식을 사용하는데, 여기에 &mu; 라는 previous direction를 나타내는 추가적인 parameter를 사용하게 된다.
Momentum을 크게 한다면 "더 멀리" 있는 공도 최적의 해로 갈 수 있도록 만들 수 있다는 것으로 이해할 수 있다.

 gradient scale에 대해 생각해 보면, gradient 의 부호 자체는 어디로 가야하는지를 잘 알려주지만 size에 대해서는 별로 효과적이지 않다. 또한 gradient의 크기가 최적화 과정에서 급격하게 변할 수 있기 때문에(경사가 급변할 수 있으므로) learning rate를 결정하는 것을 어렵게 한다.
이를 해결하기 위해 각 dimension의 gradient magnitude를 normalize 하는 방법이 있다.

### RMSProp

<img width="545" alt="image" src="https://user-images.githubusercontent.com/69765278/159140868-53265169-0980-423c-a6b1-e77564e27f41.png">

s<sub>k</sub> 는 이전의 step 값을 &beta; 를 통해 반영하여 최신 기울기가 더 크게 반영되도록 한다. 이 때 각각의 dimension 은 s의 크기로 나눠 정규화 하였다.

### AdaGrad

<img width="609" alt="image" src="https://user-images.githubusercontent.com/69765278/159140944-7db14643-a391-41b7-bfa9-b44646de3ab9.png">

RMS prop에 &beta; 가 없다. 그래서 s<sub>k</sub> 가 계속 커진다.
그렇기 때문에 learning rate &theta; 가 반복을 거듭하며 size가 작아지게 된다. 처음에는 큰 step size로 학습을 진행하고, 이후에는 step size를 작게 해 점점 세밀한 해답을 찾도록 최적화 하는 과정에 알맞아 convex problem을 푸는 데에 최적화된 알고리즘이다.
하지만 rate가 어느 정도 작아진 이후 최적해를 찾을 수 있을 때만 잘 동작한다는 단점이 있다.
따라서 RMS prop이 이를 개선한 방안이라고 생각할 수 있다.

### Adam

<img width="550" alt="image" src="https://user-images.githubusercontent.com/69765278/159141027-7aa1d86c-44b0-477b-935d-d24d7e150041.png">

momentum과 RMS prop을 섞었다.
m<sub>k</sub> 는 gradient와 이전 m의 값으로 업데이트 되며(first moment, momentum과 유사)
v<sub>k</sub> 는 gradient의 제곱과 이전 v의 값으로 업데이트 된다.(second moment)
초기 m,v의 값이 작을 경우를 대비해 m과 v에 보정을 거쳐 m hat 과 v hat을 사용하게 된다.
default 값을 보면 최적의 alpha, beta 값들이 있는데
&beta;<sub>2</sub>가&beta;<sub>1</sub> 보다 큰 이유는 b1은 momentum, b2는 magnitude에 관한 것인데, 이는 magnitude는 direction 만큼 빠르게 변하지 않기 때문이다.

# Part 3
## Stochastic gradient descent

<img width="553" alt="image" src="https://user-images.githubusercontent.com/69765278/159141250-f1c53ab0-59ba-4715-adda-67edc9942d32.png">

parameter를 업데이트 하는데 모든 데이터셋을 사용하여 loss를 구한다면 너무 시간이 오래 걸릴 것이다. 따라서 몇개의 랜덤한 sample을 통해 업데이트 하는 것을 stochastic gradient descent라고 한다.

### Stochastic gradient descent

<img width="671" alt="image" src="https://user-images.githubusercontent.com/69765278/159141268-1c387701-37e9-43d7-81d4-3c083c7d01d2.png">

<img width="419" alt="image" src="https://user-images.githubusercontent.com/69765278/159141309-4d4351ba-cda5-452a-83c3-5fd369616359.png">

하지만 항상 random하게 데이터에 access하는 것은 memory 이슈로 느리다. 따라서 dataset을 **한번** 랜덤하게 섞은 뒤 앞에서부터 정해진 수의 sample을 뽑아서 이를 수행한다.

<img width="410" alt="image" src="https://user-images.githubusercontent.com/69765278/159141336-ad86d691-315d-4893-b993-d20ffa9fc6f1.png">

빨간색은 좋은 learning rate를 골랐을 때이다.
초록색은 low learning rate를 골랐을 때이다. 이는 plateau 의 존재로 인해, worse value에 갇히는 경우로 이어질 수가 있다.

<img width="237" alt="image" src="https://user-images.githubusercontent.com/69765278/159141361-9416d21e-4946-4130-a25d-143acf415efc.png">

노란색은 high learning rate를 골랐을 때이다. 따라서 최적해로 수렴하지 못하고 벽을 타고 오르는 overshoot이 일어날 수 있다.

<img width="237" alt="image" src="https://user-images.githubusercontent.com/69765278/159141379-79131f8f-aff3-4eb2-a65e-4a1e91d08c8b.png">

따라서, learning rate를 갈수록 줄이는 방법을 사용한다.

<img width="399" alt="image" src="https://user-images.githubusercontent.com/69765278/159141397-0594dd03-1125-4b26-adbe-47b01f1eae05.png">

이를 Learning rate decay라고 하는데, SGD, momentum을 사용하거나 ADAM을 사용하는 방식으로 이를 구현한다. ADAM을 사용하는게 tuning 하기 더 쉽다고 한다.

## Tuning (stochastic) gradient descent

하이퍼파라미터는 아래와 같은 기준을 고려하여 적절히 설정할 수 있다.

<img width="568" alt="image" src="https://user-images.githubusercontent.com/69765278/159141438-abc48710-fa19-4f48-9aa8-e7b89e069a0c.png">

B, 즉 SGD에 사용되는 batch size는 클수록 좋은 값을 얻을 수 있지만 expensive하다.
learning rate는 크게 설정하고 시간에 따라 decay되어야 하며
momentum은 0.99라는 디폴트 값을 사용하는 것이 좋으며
Adam parameters 도 주어진 디폴트 값을 사용하는 것이 좋다.

hyperparameter를 대부분 training loss에서 tuning 하지만, Validation loss 단계에서 tuning하기도 한다.
